src/state/policy.rs

```diff
@@ -71,7 +71,23 @@ impl PolicyStore {
                     .or_default()
                     .insert(key.clone());
             }
-            RbacScope::WorkloadSelector => {}
+            RbacScope::WorkloadSelector => {
+                let existing = self.by_key.get(&key);
+                match existing {
+                    Some(e) => {
+                        match e.scope {
+                            RbacScope::Global => {
```

Comment:
Actually, looking at this and the surrounding code, I think maintaining multiple lists for some cases and not others is what I find really troublesome - I don't see the point in `by_namespace` if we don't also have `by_workload` or `by_global`.

It leads to this need to keep things synced in some cases but not others, which feels like a footgun. If it's intended a perf optimization, there are better ways to do sharding.


(all of that predates this fix tho, I am just looking at this code now).

Reply:
I 100% agree with the fragileness, and it applies to the rest of `state` as well. But we cannot fix that by removing all indexes and having O(N) lookups everywhere (which quickly become O(N^2) when composed).

This is exactly why we built `krt`. Not sure a good lightweight approach that solves it though

Reply:
Yep I get that - maybe better to have `by_workload` and `by_namespace` then - right now we just have `everything && by_namespace` which probably isn't going to work well for either lookups or footguns.